requirements.txt

```diff
@@ -24,3 +24,10 @@ scikit-optimize==0.5.2
 
 # Required for plotting data
 #plotly==3.1.1
+
+# find first, C search in arrays
+py_find_1st==1.1.1
```

Comment:
i'll be honest, i don't like this dependency because it doesn't install cleanly. if you happen to find a cleaner package doing the same thing ...

a small hint on the failing tests - install `future` before `pip install -r requirements.txt` in the .travis script (and dockerfile and ... )

Reply:
This is core to performance. 
Its  10's to 100s of times faster and gets used every loop twice.

There is no replacement im aware of in Python to do what this does, which is find the next match and not scan the entire array. 

The gain may not sound a lot, macro/mill to 1/10ths of seconds - but when you multiply that our by number of trades in frame, x 2 for sell / stop, X 11 different stop values * 350 markets you're talking about saving 5-10minutes

Reply:
consider a trade typically ends 10s to hundred rows later -- to scan 100,000 rows to find the next sell=1 or stop > low - its massively faster

Reply:
@xmatthias I did a quick research I didn't find any equivalent of this library. I can delve into details to see how we can replace it if it is really annoying. 
Also I am not sure if I get your hint. I didn't touch travis script. would you explain more please?

Reply:
one addition - if i understood this correctly, it finds the first value validating a equation to true in an array ...

the following 2 links seem to do explain something similar (but with np native methods) - but maybe i miss what the package really does ...

https://stackoverflow.com/questions/16243955/numpy-first-occurrence-of-value-greater-than-existing-value
https://docs.scipy.org/doc/numpy/reference/generated/numpy.argmax.html

```python
>>> import numpy as np
>>> a = np.arange(6)
>>> np.argmax(a)
5
>>> np.argmax(a > 2)
3
```

Reply:
not how i read it - to find the max it has to do a full table scan - its not finding the `next` and stopping there.

Reply:
it may find the next -after- a full scan, but by that time for binance we're already 5 minutes later be the time tested
100,000 candles X 50trades per df X twice per loop X 11 stop X 350 markets
as opposed to reading 100avg candles forwards  - monster difference in time taken

it would be like making every query in a database a full tablescan always

Reply:
yep - from the link you provided - argmax performace drops the size of the array - indicative its scanning the entire array every loop

<img width="636" alt="screen shot 2018-09-22 at 11 09 39 am" src="https://user-images.githubusercontent.com/34645187/45916586-21e73480-be58-11e8-9276-6ffeedd454a5.png">

Reply:
```
import numpy as np
import utils_find_1st as utf1st
import timeit

def s():
    st = timeit.default_timer()
    return st

def f(st):
    return (timeit.default_timer() - st)

st = s()
a = np.random.randint(1, size=999999999)
a = np.append(a, 1)
utf1st.find_1st(a, 1, utf1st.cmp_equal)
print("py_find_1st: %s" % f(st))

st = s()
b = np.random.randint(1, size=999999999)
b = np.append(b, 1)
np.argmax(b == 1)
print("argmax: %s" % f(st))
```

Reply:
timing how long it takes to generate 999999999 rows isnt part of it

Reply:
I know. it doesn't change anything as you are comparing. take it out if you want from both. same

Reply:
really not a good comparison for a test - a sell to stop will tend to be in the next hundred rows - and then there be many of them, 

IF there was only 1 trade in a massive dataset then a single-scan is going to perform relatively similar 

IF there are 50 bits only finding the next, then saving 50 full scans is where you;d see the gain

Reply:
this is more  like it - but still very broken. 
to find 50 trades find_1st only needs to scan the table once. 

to find 50 trades arg max has to scan it 50 times. 
```
import numpy as np
import utils_find_1st as utf1st
import timeit

def s():
    st = timeit.default_timer()
    return st

def f(st):
    return (timeit.default_timer() - st)


a = np.random.randint(1, size=1000000)
a = np.append(a, 1)

st = s()
for i in range(1):
    utf1st.find_1st(a, 1, utf1st.cmp_equal)
print("py_find_1st: %s" % f(st))


b = np.random.randint(1, size=1000000)
b = np.append(b, 1)

st = s()
for i in range(50):
    np.argmax(b == 1)
print("argmax: %s" % f(st))
```

```
py_find_1st: 0.0019038530008401722
argmax: 0.12748218899650965
```

Reply:
py_1st is always slightly ahead of argmax. but when developing the code, you had this idea of slapping realtime which I reckon it is not necessary in current architecture as a delay of 30 minutes or even more is enough for predicting your game. if we can test with argmax on lets say 200 pairs on 3 months and then the overall time is only 10% to 15% more than with py_1st, I would say it is maybe simpler to keep py_1st.

Reply:
X that by 500,000 as we're looking at millions of permutations - but already accounted for the 50 trades part -- its 20seconds vs over 10 minutes.

Reply:
@xmatthias would you please take a look at .travis and Dockerfile ? I added future before requirements.txt , it seems fine now but I'll appreciate your confirmation.